pageserver/page_api/proto/page_service.proto

```diff
@@ -0,0 +1,202 @@
+}
+
+// Common request fields.
+message RequestCommon {
+  // The LSN to read at. Required.
+  uint64 request_lsn = 1;
+  // If given, the caller guarantees that the page has not been modified since
+  // this LSN. Must be smaller than request_lsn. This allows the Pageserver to
+  // serve an old page without waiting for the request LSN to arrive.
+  uint64 not_modified_since_lsn = 2;
+}
+
+// A relation identifier.
+message RelTag {
+    uint32 spc_oid = 1;
+    uint32 db_oid = 2;
+    uint32 rel_number = 3;
+    uint32 fork_number = 4;
+}
+
+// Requests the size of a database, as # of bytes. This is only accurate on
+// shard 0; other shards will return their view of the database according to
+// which pages they have.
+message DbSizeRequest {
+  RequestCommon common = 1;
+  uint32 db_oid = 2;
+}
+
+message DbSizeResponse {
+  uint64 num_bytes = 1;
+}
+
+// Requests a base backup at a given LSN.
+message GetBaseBackupRequest {
+  // The LSN to fetch a base backup at.
+  RequestCommon common = 1;
+  // If true, logical replication slots will not be created.
+  bool replica = 2;
+}
+
+// Base backup response chunk, returned as an ordered stream.
+message GetBaseBackupResponseChunk {
+  // A basebackup data chunk. The size is undefined, but bounded by the 4 MB
+  // gRPC message size limit.
+  bytes chunk = 1;
+}
+
+// Requests a single page.
+message GetPageRequest {
+  // A request ID. Will be included in the response. Should be unique for
+  // in-flight requests on the stream.
+  uint64 id = 1;
+  // The LSN to read at.
+  RequestCommon common = 2;
+  // The relation to read from.
+  RelTag rel = 3;
+  // The page number to read. Must belong to the remote shard.
+  uint32 block_number = 4;
+  // The request class.
+  GetPageClass class = 5;
+}
+
+// A GetPageRequest class. Primarily intended for observability, but may also be
+// used for prioritization in the future.
+enum GetPageClass {
+  // Unknown class. For forwards compatibility: used when the client sends a
+  // class that the server doesn't know about.
+  GET_PAGE_CLASS_UNKNOWN = 0;
+  // A normal request. This is the default.
+  GET_PAGE_CLASS_NORMAL = 1;
+  // A prefetch request.
+  GET_PAGE_CLASS_PREFETCH = 2;
+}
+
+// A batch of GetPage requests. These will be executed as a single batch by the
+// Pageserver, amortizing layer access costs and parallelizing them. This may
+// increase the latency of any individual request, but improves the overall
+// latency and throughput of the batch as a whole.
+//
+// Responses will be emitted individually, as soon as they are ready. They may
+// be emitted in a different order than the requests.
+message GetPageRequestBatch {
+  repeated GetPageRequest requests = 1;
+}
+
+// A GetPage response. May be emitted out of order.
+message GetPageResponse {
+  // The original request's ID.
+  uint64 id = 1;
+  // The response status code.
+  GetPageStatus status = 2;
+  // A string describing the status, if any.
+  optional string reason = 3;
```

Comment:
nit: not sure why we need optional here. It's a little bit inconsistent that `reason` is optional, but `page_image` is not.

Another possibility is to use `oneof` instead of status enum (`oneof` is more like rust's enums) to underline that these fields are mutually exclusive and are tied to status

Reply:
I'll just drop the `optional`. The `optional` generates an `Option<String>` in Rust, but it doesn't matter since empty strings won't be heap-allocated anyway, and we're going to convert these into Rust domain types in a later PR anyway (so we can do whatever preprocessing we want there).

> Another possibility is to use oneof instead of status enum (oneof is more like rust's enums) to underline that these fields are mutually exclusive and are tied to status

That would require adding a separate message kind for every status, where most kinds would also have a `reason` field. Seems a bit overengineered.